pulp_rpm/app/models/repository.py

```diff
@@ -350,6 +359,20 @@ def finalize_new_version(self, new_version):
                     "{value_errors}"
                 ).format(repo=new_version.repository.name, value_errors=str(ve))
             )
+    
+    def _check_signatures(self, new_version):
+        """If requested, validate that the added rpms are signed appropriately."""
+        if self.allowed_pub_keys:
+            rejected_nevras = []
+            for package in Package.objects.filter(pk__in=new_version.added()).iterator():
+                if str(package.signer_key_id) not in self.allowed_pub_keys:
```

Comment:
Existing content is not checked, only added packages, as seen in the snippet you quoted. This seems like the least surprising way to handle this restriction, but if there are better options I'm all ears. Perhaps the `RepositoryVersion` should copy and make immutable the list of pubkeys that were allowed when it was created? I don't know what the purpose of that would be though. You can't go back and edit the content of a previous RepoVersion can you? Just create a new one. So the only thing that matters is what the pub key list is _now_.

Reply:
I don't understand what you're trying to say about moving the check to upload/sync time. Adding content to a repo and syncing both of course create a new repo version, so this check does currently run at sync/content adding time.

Reply:
here is a check added to ansible at upload time https://github.com/pulp/pulp_ansible/blob/5ae04f630a5c1ed679f7395323dcf0cf94656c97/pulp_ansible/app/tasks/signature.py#L52
and for the sync i was thinking to right away raise error here https://github.com/pulp/pulp_rpm/pull/2954/files#r1105711127

Reply:
We are in fact checking the signature at upload time. But different repos might allow different (or any) signatures, so you have to check that the rpm's signature matches what the repo allows at repo-add time.

Also [see discussion here](https://discourse.pulpproject.org/t/enforce-package-deb-and-rpm-signatures/511/12?u=stephen_herr) about why doing the equivalent of `gpg --verify` is hard for RPMs. It would require that the `rpm`/`rpmkeys` binary is installed, which is not currently a given.

Reply:
I guess I don't object to having a second sync-only location where the repo checks that the signatures are allowed, it just seems redundant to me. Does that solve a problem?

Reply:
I understand what you are saying now in the other comment about "failing fast". I have no opinion here, except that I think if there's one location that you can do a check for every scenario that's probably the right place, vs having the same check being done at different places in different scenarios.

Reply:
@sdherr to be clear, these question are not strictly related to the rpm plugin, but pretty much every plugin that would like to add package verification would face them, so i am bringing them now for discussion

Reply:
Repo version is immutable, correct. However when you inspect previous versions, how can you decide whether they are still compliant with the rules you have in place right now?Maybe you no longer trust a key but the older version is being distributed. Compliance rule can change and practically you can rely only on the latest version distribution only.
Maybe we could take advantage of this info field https://github.com/pulp/pulpcore/issues/2998

Reply:
>It would require that the rpm/rpmkeys binary is installed, which is not currently a given.

Well the library certainly has to be installed otherwise createrepo_c won't be able to read the packages either.  But maybe debian packages the library and executables separately, haven't looked into it.

> Or we could simply add an endpoint to audit against list of allowed keys set on the repo

Long term my feeling is that allowed GPG keys is not going to be the only criteria that users may want to have, so perhaps we need some way to define generic "compliance rules" for a repo?  However that is too big a task to chew on currently.

I can imagine a use case for both scenarios where you want to allow older versions to have one set of compliance rules while changing them over time, and having all available versions follow one set of compliance rules, so this is a difficult problem.

In practice though (and also based on some user research that was done), we don't expect *most* users to care about more than the last, say, 5 repository versions.  They want the latest and then the last few versions to potentially roll back to if needed.  So I don't want to overthink this too much.

Reply:
>     I don't believe that's true. I think createrepo_c re-implements a lot of information about the RPM file structure in order to avoid that dependency. If I'm wrong let me know, because if rpmlib is installed then there are a lot more options here.

It has RPM as a hard dependency.  There are only certain bits of functionality that it doesn't rely on librpm for to increase flexibility, but the basic file format processing it does.

Reply:
It's not as obvious to me as I thought it would be how we could take advantage of that to drop the `rpmfile` dependency, but if you have a suggestion feel free to mention it.

Reply:
The same functionality should be available, though given their docs are offline it's probably hard to tell.  I'm trying to prod them to fix that currently

https://github.com/rpm-software-management/rpm-web/issues/34

Reply:
@sdherr Could you try out this implementation? https://github.com/rpm-software-management/createrepo_c/issues/346#issuecomment-1432656618

Reply:
That works if `python3-rpm` is installed, yes, but that's not a given.